lib/react.dart

```diff
@@ -419,13 +502,24 @@ abstract class Component2 implements Component {
   ///
   /// Optionally accepts a [callback] that gets called after the component updates.
   ///
-  /// Also allows [newState] to be used as a transactional `setState` callback.
+  /// To use a transactional `setState` callback, check out [setStateTransaction].
   ///
-  /// See: <https://facebook.github.io/react/docs/react-component.html#setstate>
-  void setState(dynamic newState, [SetStateCallback callback]) {
+  /// See: <https://reactjs.org/docs/react-component.html#setstate>
+  void setState(Map newState, [SetStateCallback callback]) {
     adapter.setState(newState, callback);
   }
 
+  /// Triggers a rerender with new state obtained by shallow-merging
+  /// the return value of [stateUpdater] into the current [state].
+  ///
+  /// Optionally accepts a [callback] that gets called after the component updates.
+  ///
+  /// See: <https://reactjs.org/docs/react-component.html#setstate>
+  void setStateTransaction(TransactionalSetStateCallback stateUpdater,
+      [SetStateCallback callback]) {
```

Comment:
We don't have to land on the final name right now, but it looks like the word "transaction" isn't ever used in the react docs: https://reactjs.org/docs/react-component.html#setstate ; they use "updater" instead:
> ```
> setState(updater[, callback])
>```


Could we update the naming to match? Might also be nice to have inline function types, since typedefs are harder to make sense of in the IDE:
```dart
void setStateWithUpdater(Map updater(Map state, Map props), [void callback()]);
```